### ISSUE TITLE
Create a wrapper for `Word`

### ISSUE THREAD

**Issue by vlopes11** [0xPolygonMiden/crypto#59]
Currently `Word` is a type alias

https://github.com/0xPolygonMiden/crypto/blob/0c242d2c518663f1fb0208fab7147db259b2e85e/src/lib.rs#L24-L34

However, it is a critical component of our stack as it is used everywhere, in different forms. One example is `RpoDigest`:

https://github.com/0xPolygonMiden/crypto/blob/0c242d2c518663f1fb0208fab7147db259b2e85e/src/hash/rpo/digest.rs#L7-L11

They both have the same structure, but different concrete types. This leads to many cases where unsafe optimizations are required, such as in here:

https://github.com/0xPolygonMiden/crypto/pull/58

One of the problems highlighted by this PR is the need to convert between `Word` and `RpoDigest` efficiently (i.e. without copying/allocating).

We could achieve major simplification if:

- `Word` was a struct
- The digest of `Rpo256` was a `Word`

**Comment by vlopes11** [comment]
@bobbinth should we include this on 0.6? It will clean the code quite a bit as we won't have to convert between the types anymore

**Comment by bobbinth** [comment]
Yes - let's do it!

**Comment by sergerad** [comment]
@bobbinth this is a really old issue. From what I can see it was never completed given this is still in next:
```rust
pub type Word = [Felt; WORD_SIZE];
...
pub struct RpoDigest([Felt; DIGEST_SIZE]);
```

Do we still want to do it? Any changes or additions to the issue you would make?

**Comment by bobbinth** [comment]
I'm actually not too sure what's the best way forward on this. One option is to just add a newtype for word which could look like so:
```rust
pub struct Word([Felt; DIGEST_SIZE]); 
```
But then it is basically identical to `RpoDigest` and the question is whether we need to keep both.

Maybe what we need to do here is to get rid of `RpoDigest` (and `RpxDigest` as well), and have just the `Word` newtype. But we should probably evaluate if that could cause any issue (I don't think it should, but it's also been a while since I've been deep in this code).

**Comment by sergerad** [comment]
Thanks, yea I'll have a poke around and see what makes sense.

**Comment by sergerad** [comment]
Thats good to know. Need to decide what to do with `RpxDigest`. `Word` might have to replace both `RpxDigest ` and `RpoDigest` I suppose.

**Comment by plafer** [comment]
Why would removing `RpoDigest` be desirable? I assume it's the copious amounts of `.into()` between `Word` and `RpoDigest`? While more convenient, technically marking things as needing an `RpoDigest` is more descriptive than just a `Word`. For example, if/when we switch to using `RpxDigest`, we will get nice compiler errors for anywhere where we pass in an `RpxDigest` but that still expects an `RpoDigest`. If we just used `Word` everywhere, we wouldn't get any compiler errors.

**Comment by bobbinth** [comment]
> I presume `RpoDigest` and `RpxDigest` have domain-specific functionality that wouldn't make sense for `Word` to contain (otherwise there would only be 1 digest type?).

I actually don't think there is that much specific functionality and they probably can be the same type.

> Why would removing `RpoDigest` be desirable? I assume it's the copious amounts of `.into()` between `Word` and `RpoDigest`? While more convenient, technically marking things as needing an `RpoDigest` is more descriptive than just a `Word`. For example, if/when we switch to using `RpxDigest`, we will get nice compiler errors for anywhere where we pass in an `RpxDigest` but that still expects an `RpoDigest`. If we just used `Word` everywhere, we wouldn't get any compiler errors.

This could be a good reason to keep them separate. But I could also see that if the digests are meaningful somehow, they should have more narrow newtypes (e.g., `Nullifier(Word)`). Another point is that we frequently use `RpoDigest` and `Word` interchangeably (e.g., in Miden base storage slots are words but some of them are roots of Merkle trees which are digests) - so, we frequently break this "type safety".

**Comment by PhilippGackstatter** [comment]
> But I could also see that if the digests are meaningful somehow, they should have more narrow newtypes

I agree. We've previously discussed introducing newtypes for type safety such as `BlockCommitment(Digest)`, `AccountRoot(Digest)`, etc. So very narrowly defined, rather than having a more generic "digest". I wonder if this would work in other places too?

**Comment by plafer** [comment]
Yeah that makes sense to me - so probably makes sense to remove `RpoDigest` in that case.

**Comment by bobbinth** [comment]
If we get rid of `RpoDigest` we probably don't need `RpxDigest` either.

**Comment by sergerad** [comment]
I have started chipping away at this in the above PR by
- Renaming Word->OldWord
- Renaming RpoDigest->Word
- Renaming OldWord->Word and dealing with compilation errors

There are a huge amount of diffs but so far it doesn't seem like there is anything too complicated. Its just laborious. I have dealt with ~100 errors from the origin ~200 so far.

**Comment by PhilippGackstatter** [comment]
One thing that I think we'll need in miden-base is a macro that auto-generates a newtype wrapper of `Word`. We have probably a dozen types where we would want to create wrappers for. Creating and maintaining all of those types by hand seems like a bad idea. But perhaps there is a similar need in crypto or VM, so we could consider adding such a macro in the crypto crate directly. This could be a separate issue/PR of course.

I haven't thought too much about it, but a proc macro might be the most flexible. For example:

```rust
#[derive(Debug, Clone, WordWrapper)]
pub struct VaultRoot(Word);
```

That `WordWrapper` macro would re-export all the `Word` (previously `Digest`) methods like `as_elements`, `to_hex`, etc and implement standard traits. For each type, we could still add our own methods if that is somehow necessary, which wouldn't be possible with a declarative macro afaict.

**Comment by sergerad** [comment]
Note we have this https://github.com/0xMiden/miden-base/issues/1373

**Comment by sergerad** [comment]
Ah sorry I had missed this and thought you were referring to the other macro.
> One thing that I think we'll need in miden-base is a macro that auto-generates a newtype wrapper of Word.
